### PR TITLE
[For discussion / feedback] option to crop only content

### PR DESCRIPTION
**Pull request by untimished**
Option to clip the content instead of cropping the entire page.
The poppler preview displays the regular cropping method, but saving as pdf retains the page, clipping only the content.

Issue: #471

dialogue

Rotating pages causes issues and needs to be fixed.
This should be a draft. Excluding changing the internal preview means that this option can be implemented with very few changes.

### CODE REVIEW

**Comment by angsch** [comment]
Looks promising. I don't know how to deal with the rendering issue and hope that someone else can comment on it.

### Crops are not rotated on rotated pages
The crops have to be rotated on export. 
1. Load a page and rotate it to the right
2. Set right crop to 20% and tick the checkbox
3. Export

The exported PDF has the wrong format.
image

### crop white borders
The current code for crop white borders is not compatible.
1. Load a page
2. Crop white borders

I get the error message:
`TypeError: crop() missing 1 required positional argument: 'clip'`
Perhaps we can make the clip argument optional?

### Select Same Page Format
This selection is based on the rendered result. Hence, real crop and the new feature are considered to have the same page format.

1. Load 2 pages.
2. Apply some crop to the first page
3. Apply the same crop to the second page and tick the checkbox.
4. Select Same Page Format

Both pages are selected although they do not have the same page format.

**Comment by kbengs** [comment]
Yes, looks promising.
In is_data_valid the test len(tmp) == 9 need to be 10 now.

I did a test with the rendering. It's a start but there are more things to sort out.
```
--- a/pdfarranger/iconview.py
+++ b/pdfarranger/iconview.py
@@ -48,6 +48,11 @@ class CellRendererImage(Gtk.CellRenderer):
             return
 
         w0, h0, w1, h1, w2, h2, rotation = self.get_geometry()
+        if self.page.clip:
+            if w1 == w0:
+                w2, h2 = w0, h0
+            else:
+                w2, h2 = w1, h1
         th = int(2 * self.th1 + self.th2)
         w = w2 + th
         h = h2 + th
@@ -69,17 +74,37 @@ class CellRendererImage(Gtk.CellRenderer):
         window.rectangle(0, 0, w2 + 2 * self.th1, h2 + 2 * self.th1)
         window.fill()
 
-        # image
+        # white paper
         window.set_source_rgb(1, 1, 1)
         window.rectangle(self.th1, self.th1, w2, h2)
+        window.fill()
+
+        # image
+        window.set_source_rgb(1, 1, 1)
+        if self.page.clip:
+            window.translate(self.th1, self.th1)
+            a = self.page.crop[0] * w1
+            b = self.page.crop[2] * h1
+            c = w1 - self.page.crop[0] * w1 - self.page.crop[1] * w1
+            d = h1 - self.page.crop[2] * h1 - self.page.crop[3] * h1
+        else:
+            a = self.th1
+            b = self.th1
+            c = w2
+            d = h2
+        window.rectangle(a, b, c, d)
         window.fill_preserve()
         window.clip()
 
-        window.translate(self.th1, self.th1)
+        if not self.page.clip:
+            window.translate(self.th1, self.th1)
         scale = self.page.resample * self.page.zoom
         window.scale(scale, scale)
         x = int(self.page.crop[0] * w1)
         y = int(self.page.crop[2] * h1)
+        if self.page.clip:
+            x = 0
+            y = 0
         window.translate(-x, -y)
         if rotation > 0:
             window.translate(w1 / 2, h1 / 2)
```

File: pdfarranger/core.py [hunk 1]
```diff
@@ -319,8 +321,9 @@ def move(self, treerowref, before):
         self.before = before
         self.treerowref = treerowref
 
-    def addpages(self, filename, page=-1, basename=None, angle=0, scale=1.0, crop=None):
+    def addpages(self, filename, page=-1, basename=None, angle=0, scale=1.0, crop=None, clip=None):
```
Comment (kbengs) [inline pdfarranger/core.py:324]: Could you have clip=False here and skip the next line?

**Comment by untimished** [comment]
Really appreciate the help, thank you for your comments and proposals.
@kbengs I will **focus on exporting first** and then use your diff for the **preview display later on**.

There is a **big remaining issue**, with **copying cropped/clipped files** - these get **cropped repeatedly or shrink**.

File: pdfarranger/exporter.py [hunk 2]
```diff
@@ -147,13 +147,12 @@ def export(input_files, pages, file_out, mode, mdata):
         angle = row.angle
         angle0 = current_page.Rotate if '/Rotate' in current_page else 0
         new_page = pdf_output.copy_foreign(current_page)
-        #new_page.Rotate = 0
+        new_page.Rotate = 0
```
Comment (untimished) [inline pdfarranger/exporter.py:150]: Setting new_page.Rotate = 0 and adding an optional angle argument to  _mediabox() in exporter.py is changing quite a lot and a bit of a hack.

**Comment by kbengs** [comment]
> And copy_foreign Annotations throws Error:

This should be fixed with either of m-holgers PRs.

**Comment by mara004** [comment]
As a design consideration, I think it might be better to make trimming pages a separate feature, for one might want to apply both crop and trim.

File: pdfarranger/exporter.py [hunk 3]
```diff
@@ -146,9 +144,25 @@ def export(input_files, pages, file_out, mode, mdata):
...
-            new_page.Rotate = angle + angle0
-        new_page.MediaBox = _mediabox(new_page, row.crop)
+        new_page.Rotate = 0
+        mediabox = _mediabox(new_page, row.crop, angle + angle0)
+        if not row.clip:
+            new_page.MediaBox = mediabox
+        # clip content instead of cropping the page
+        else:
+            print(row.crop)
+            new_page.CropBox = _mediabox(new_page, row.crop, angle + angle0)
```
Comment (mara004) [inline pdfarranger/exporter.py:154]: You could alternatively use `TrimBox`

**Comment by kbengs** [comment]
"Hide margins" was implemented in https://github.com/pdfarranger/pdfarranger/pull/954

**Pull request closed without merging**